stackslib/src/chainstate/nakamoto/mod.rs

```diff
@@ -422,47 +431,84 @@ impl NakamotoBlockHeader {
             state_index_root: TrieHash([0u8; 32]),
             miner_signature: MessageSignature::empty(),
-            stacker_signature: MessageSignature::empty()
+            stacker_signature: MessageSignature::empty(),
         }
     }
 }
 
 impl NakamotoBlock {
-    /// Did the stacks tenure change on this nakamoto block? i.e., does this block
-    ///  include a TenureChange transaction?
-    pub fn tenure_changed(&self, parent: &StacksBlockId) -> bool {
+    /// Find all positionally-valid tenure changes in this block.
+    /// They must be the first transactions.
+    /// Return their indexes into self.txs
+    fn find_tenure_changes(&self) -> Vec<usize> {
+        let mut ret = vec![];
+        for (i, tx) in self.txs.iter().enumerate() {
+            if let TransactionPayload::TenureChange(..) = &tx.payload {
+                ret.push(i);
+            } else {
+                break;
+            }
+        }
+        ret
+    }
+
+    /// Does this block contain one or more well-formed and valid tenure change transactions?
+    /// Return Some(true) if it does contain at least one, and they're all valid
+    /// Return Some(false) if it does contain at least one, but at least one is invalid
+    /// Return None if it contains none.
+    pub fn tenure_changed(&self) -> Option<bool> {
+        let wellformed = self.is_wellformed_first_tenure_block();
+        if wellformed.is_none() {
+            // block isn't a first-tenure block, so no valid tenure changes
+            return None;
+        } else if let Some(false) = wellformed {
+            // this block is malformed
+            info!("Block is malformed";
+                  "block_id" => %self.block_id());
+            return Some(false);
+        }
+
         // Find all txs that have TenureChange payload
         let tenure_changes = self
-            .txs
+            .find_tenure_changes()
             .iter()
-            .filter_map(|tx| match &tx.payload {
-                TransactionPayload::TenureChange(payload) => Some(payload),
-                _ => None,
-            })
+            .map(|i| &self.txs[*i])
             .collect::<Vec<_>>();
 
         if tenure_changes.len() > 1 {
-            warn!(
+            debug!(
                 "Block contains multiple TenureChange transactions";
                 "tenure_change_txs" => tenure_changes.len(),
                 "parent_block_id" => %self.header.parent_block_id,
                 "consensus_hash" => %self.header.consensus_hash,
             );
         }
 
-        let validate = |tc: &TenureChangePayload| -> Result<(), TenureChangeError> {
-            if tc.previous_tenure_end != *parent {
-                return Err(TenureChangeError::PreviousTenureInvalid);
-            }
+        let validate = |tc: &StacksTransaction| -> Result<(), TenureChangeError> {
+            if let TransactionPayload::TenureChange(tc) = &tc.payload {
+                if tc.previous_tenure_end != self.header.parent_block_id {
```

Comment:
What does it mean if there are multiple tenure change transactions in a block? Would all of those tenure change transactions have the same `previous_tenure_end` field?

Reply:
There can be N > 1 tenure changes in block if and only if the past N - 1 tenures did not contain any blocks.  These are empty tenures, so their parent blocks are necessarily the same as the chain tip atop which the Nth tenure change builds if this block is valid.

Reply:
Put another way, the `previous_tenure_end` points to the chain tip at which this tenure begins.  This is true even if the tenure is empty.  If N tenures in a row are empty, then their N tenure change transactions would have the same value for `parent_tenure_end` (since the chain tip is the same in all cases).  The N+1'st tenure's tenure change transaction would also have the same `parent_tenure_end` value.

Reply:
Should point out -- this PR is not trying to handle the case where we need to issue a tenure change to deal with an empty sortition or a tenure budget extension.  I didn't think that was in-scope.  Right now, this PR only considers a tenure change to be valid if it is signaling the beginning of a new miner tenure triggered by a winning sortition.

Reply:
I suspect this code, as well as `is_wellformed_first_tenure_block()`, is going to need to be modified in the follow-on PR that fully validates tenure-changes.

Reply:
Okay, that makes sense to me then.